kafka-client/src/test/java/dev/responsive/kafka/internal/db/CassandraFactTableIntegrationTest.java

```diff
@@ -141,57 +150,225 @@ public void shouldUseTwcsWithoutTtl() throws Exception {
         .getTable(tableName)
         .get();
     final String describe = table.describe(false);
-    assertThat(describe,
-        containsString("'org.apache.cassandra.db.compaction.TimeWindowCompactionStrategy'"));
+
+    assertThat(describe, containsString("default_time_to_live = " + (int) defaultTtl.toSeconds()));
   }
 
   @Test
-  public void shouldInsertAndDelete() throws Exception {
+  public void shouldRespectSemanticKeyBasedTtl() throws Exception {
     // Given:
-    params = ResponsiveKeyValueParams.fact(storeName);
-    final RemoteKVTable<BoundStatement> table = client
-        .factFactory()
-        .create(RemoteTableSpecFactory.fromKVParams(params, defaultPartitioner()));
+    final var defaultTtl = Duration.ofMinutes(30);
+    final Function<String, Optional<TtlDuration>> ttlForKey = k -> {
+      if (k.equals("KEEP_FOREVER")) {
+        return Optional.of(TtlDuration.noTtl());
+      } else if (k.endsWith("DEFAULT_TTL")) {
+        return Optional.empty();
+      } else {
+        final long ttlMinutes = Long.parseLong(k);
+        return Optional.of(TtlDuration.of(Duration.ofMinutes(ttlMinutes)));
+      }
+    };
+    final TtlProvider<String, String> ttlProvider = TtlProvider
+        .<String, String>withDefault(defaultTtl)
+        .fromKey(ttlForKey, Serdes.String());
+    params = ResponsiveKeyValueParams.fact(storeName).withTtlProvider(ttlProvider);
+
+    final var table = client.factFactory().create(RemoteTableSpecFactory.fromKVParams(
+        params,
+        defaultPartitioner(),
+        Optional.of(new TtlResolver<>(false, "changelog-ignored", ttlProvider))
+    ));
 
     table.init(1);
 
-    final Bytes key = Bytes.wrap(new byte[]{0});
+    final Bytes noTtlKey = serializedKey("KEEP_FOREVER");
+    final Bytes defaultTtlKey = serializedKey("DEFAULT_TTL"); // default is 30min
+    final Bytes tenMinTtlKey = serializedKey(String.valueOf(10L));
+    final Bytes fiftyMinTtlKey = serializedKey(String.valueOf(50L));
+
     final byte[] val = new byte[]{1};
 
     // When:
-    client.execute(table.insert(1, key, val, CURRENT_TIME));
-    final byte[] valAt0 = table.get(1, key, MIN_VALID_TS);
-    client.execute(table.delete(1, key));
-    final byte[] valAt1 = table.get(1, key, MIN_VALID_TS);
+    final long insertTimeMs = 0L;
+    client.execute(table.insert(1, noTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, defaultTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, tenMinTtlKey, val, insertTimeMs));
+    client.execute(table.insert(1, fiftyMinTtlKey, val, insertTimeMs));
 
     // Then
-    assertThat(valAt0, is(val));
-    assertThat(valAt1, nullValue());
+    long lookupTime = Duration.ofMinutes(11).toMillis();
```

Comment:
no need to change it for this PR, but just in the future you can use `TimeUnit.MINUTES.toMillis(11)` etc.. for all of these. IMO a little easier to read and has no object allocations 😅